include/mata/synchronized_iterator.hh

```diff
@@ -0,0 +1,260 @@
+                this->positions.push_back(vector.begin());
+                this->ends.push_back(vector.end());
+            };
+
+            // Empties positions and ends.
+            // Though they should keep the allocated space.
+            void reset(const int size = 0) {
+                positions.clear();
+                ends.clear();
+                if (size > 0) {
+                    positions.reserve(size);
+                    ends.reserve(size);
+                }
+            };
+
+            virtual bool advance() = 0;
+            virtual const std::vector<Iterator> get_current() = 0;
+        };
+
+
+
+        template<typename Key>
+        class SynchronizedUniverzalIterator: public SynchronizedIterator<Key> {
+        public:
+
+            using Iterator = typename OrdVector<Key>::const_iterator;
+
+            // "minimum" would be the smallest class bounded from below by all positions that appears in all OrdVectos.
+            // Are all positions at this class?
+            // Invariant: it can be true only if all positions are indeed synchronized.
+            bool synchronized_at_current_minimum = false;
+
+            // advance() advances all positions to the NEXT minimum and returns true,
+            // or returns false if positions cannot be synchronized.
+            // If positions are synchronized to start with,
+            // then synchronized_at_current_minimum decides whether to stay or advance further.
+            // The general of the algorithm is to synchronize everybody with position[0].
+            bool advance()
+            {
+                //Nothing to synchronize.
+                if (this->positions.empty())
+                    return false; //?? or not?
+
+                // If already synchronized, start moving forward by advancing position[0] (and so break synchronization).
+                if (this->synchronized_at_current_minimum) {
+                    this->positions[0]++;
+                    synchronized_at_current_minimum = false;
+                }
+
+                // If positions[0] has nowhere to go, then sync is not possible.
+                if (this->positions[0] == this->ends[0])
+                    return false;
+
+                // Synchronise all positions with position[0].
+                for (size_t i = 1; i < this->positions.size(); i++)
+                {
+                    // If some positions has nowhere to go, then sync is not possible.
+                    if (this->positions[i] == this->ends[i])
+                        return false;
+
+                    //  Advance position[i] and position[0] to the closest equal values.
+                    while (*this->positions[i] != *this->positions[0]) {
+
+                        // Advance position[i] to or beyond position[0].
+                        while (*this->positions[i] < *this->positions[0]) {
+                            this->positions[i]++;
```

Comment:
Again, I would prefer using the `++i` version. The latter might create unnecessary object.